sympy/core/operations.py

```diff
@@ -504,3 +505,129 @@ def args(self):
+    """
+    Handler dispatcher for associative operators
+
+    .. notes::
+       This approach is experimental, and can be replaced or deleted in the future.
+       See https://github.com/sympy/sympy/pull/19463.
+
+    Explanation
+    ===========
+
+    If a new class is meant to return the result other than `Add` or `Mul` when operated,
+    it should have its handler defined as class method. Also, if the arguments have the same
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if issubclass(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    See Also
+    ========
+
+    core.compatibility.OpDispatcher
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.name = name
+        self.fallback = fallback
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self.dispatcher = Dispatcher(name)
+
+    def __str__(self):
+        return "<dispatched %s>" % self.name
+    __repr__ = __str__
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
+        """
+        def _(func):
+            self.add((cls1, cls2), func)
+            return func
+        return _
+
+    def add(self, signature, func, on_ambiguity=(lambda *args, **kwargs: None)):
+        # This suppresses unnecessary ambiguity warning, since not all sympy objects
+        # need to operate with each other.
+        return self.dispatcher.add(signature, func, on_ambiguity)
+
+    def __call__(self, *args, **kwargs):
+        return self.find_handler(args)(*args, **kwargs)
+
+    def find_handler(self, args):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        args_data = []
+        for arg in args:
+            handler = getattr(arg, self.handlermethod, None)
+            if handler is None:
+                continue
+            func, op_priority = type(arg), getattr(arg, '_op_priority', 0)
+            args_data.append((func, op_priority, handler))
```

Comment:
Let's make the API look more like `Dispatcher`:

```suggestion
    def __call__(self, *args, **kwargs):
        types = tuple([type(arg) for arg in args])
        return self.self.dispatch(*types)(*args, **kwargs)

    def dispatch(self, types):
        """
        Select the argument with highest priority, and return its handler.
        """
        args_data = []
        for typ in types:
            handler = getattr(typ, self.handlermethod, None)
            if handler is None:
                continue
            op_priority = getattr(typ, '_op_priority', 0)
            args_data.append((typ, op_priority, handler))
```

Reply:
This also has the effect of requiring `_op_priority` and `handler` be class / static methods, which is probably a good thing.

Reply:
Sounds nice